File: javascript/programmers/javascript/level2/파일명정렬.js
const files = [
  "img123456712341234123412341233412341234123412341234asdgsadgasdgasdg.png",
  "img103abd4567.png",
  "img023abdc4567.png",
  "img134abdc567.png",
  "IMG013abcd4567.GIF",
  "img234abdc567.JPG",
];

const solution = (files) => {
  let answer = [];
  const arr = [];
  for (let i = 0; i < files.length; i++) {
    const reg = /(\d+)/g;
    const value = files[i].split(reg);
    const file = files[i];
    const head = value[0].toUpperCase();
    let number = value[1];
    if (number.length > 5) {
      number = number.substring(0, 5);
    }
    number = Number(number);
    arr.push({ file: file, head: head, number: number, idx: i });
  }
  //정렬
  arr.sort((a, b) => {
    if (a.head < b.head) return -1;
    else if (a.head > b.head) return 1;
    else {
      if (a.number > b.number) return 1;
      else if (a.number < b.number) return -1;
      else {
        //HEAD, NUMBER가 같은 경우 index로 순서보장
        if (a.idx < b.idx) return -1;
        else return 1;
      }
    }
  });
  //답 구하기
  for (let i = 0; i < arr.length; i++) {
    answer.push(arr[i].file);
  }
  return answer;
};

console.log(solution(files));

File: python/programmers/level2/파일명정렬.py
def solution(files):
    answer = []
    arr = []
    for file in files:
        # 숫자의 시작인덱스와 끝 인덱스
        # 파싱해주기 위해 숫자의 처음인덱스와 끝인덱스를 찾는다.
        idx = 0
        nameLen = len(file)
        startIdx = None
        endIdx = None
        while idx < nameLen:
            if '0' <= file[idx] <= '9':
                endIdx = idx
                if startIdx is None:
                    startIdx = idx
            elif startIdx is not None:
                break
            if endIdx is not None and startIdx is not None and endIdx - startIdx == 4:
                break
            idx += 1

        head = file[:startIdx].lower()
        number = int(file[startIdx:endIdx + 1])
        tail = file[endIdx + 1:]
        # print("%s, %s, %s" % (head, number, tail))
        # 배열에 추가
        arr.append([file, head, number, tail])

    # 배열 정렬
    arr.sort(key=lambda x: (x[1], x[2]))

    # 정렬을 토대로 answer에 추가
    for item in arr:
        answer.append(item[0])
    return answer


print(solution(['img12.png', 'img10.png', 'img02.png', 'img1.png', 'IMG01.GIF', 'img2.JPG']))
